### ISSUE TITLE
Cannot read properties of undefined (reading 'bind') in useQuery hook

### ISSUE THREAD

**Issue by PrakashC1** [apollographql/apollo-client#11644]
Hi All, I am using graphql client for query and subscription in react web application.  I  am creating global graphqlclient object in one of my library. And one of my MFE(Microfrontend app) consuming this library and making use of this graphql client to query and subscriptions.

In that MFE, I tried to use useQuery hook for pagination purpose, But I am getting this below error in devtools. Not able to get rid of this error. Could you please anyone help me on this?

login_vendors.2c3e8563ae4cd163c206.chunk.js:2 TypeError: Cannot read properties of undefined (reading 'bind')
    at useQuery.js:369:1
    at Object.useMemo (login_vendors.2c3e8563ae4cd163c206.chunk.js:2:5864335)
    at t.useMemo (login_vendors.2c3e8563ae4cd163c206.chunk.js:2:6023820)
    at InternalState.useObservableQuery (useQuery.js:367:31)
    at InternalState.useQuery (useQuery.js:162:1)
    at useQuery (useQuery.js:50:44)
    at ManageMembers (ManageMembers.tsx:67:33)
    at Ji (login_vendors.2c3e8563ae4cd163c206.chunk.js:2:5859483)
    at _l (login_vendors.2c3e8563ae4cd163c206.chunk.js:2:5905529)
    at Ss (login_vendors.2c3e8563ae4cd163c206.chunk.js:2:5898289
)

Library pakage.json
---------------------
"apollo-cache-inmemory": "^1.6.5",
    "apollo-client": "^2.6.8",
    "apollo-link": "^1.2.13",
    "apollo-link-error": "^1.1.12",
    "apollo-link-http": "^1.5.16",
    "apollo-link-ws": "^1.0.19",
    "apollo-utilities": "^1.3.3",
    "graphql": "^16.8.1",
    "graphql-tag": "^2.12.6",
    "subscriptions-transport-ws": "^0.9.16",
.....

MFE pakage.json
------------------
 "@apollo/client": "^3.9.5",
"graphql": "^16.8.1",
......

Thanks in advance......

**Comment by phryneas** [comment]
Could you explain the "library package.json" vs "MFE package.json" discrepancy a bit here?

Apollo Client 2 and 3 are not really compatible with each other, so if in this case the `useQuery` hook (AC3) would try to interact with an AC2 instance, that would explain your breakage.

**Comment by PrakashC1** [comment]
HI @phryneas,
Sorry due to some reasons, I can't share my code here. I will try to explain as much as I can.

I have created a custom npm library for my project that includes a function to create a GraphQL client. This library is designed to be a singleton, meaning it's intended to be instantiated only once and reused across your project. This is achieved using the webpack-module-federation plugin.

The custom npm library, let's call it `CustomLib`, is version 1.5.0. It exposes a function to create a GraphQL client, which is then stored in a global variable.

You have two Micro Frontends (MFEs), `HostMFE` and `AdminMFE`.

`HostMFE` uses `CustomLib` as a singleton library. It creates a GraphQL client using `CustomLib` and also imports some React components from `AdminMFE`. Since `HostMFE` is the first to load, it's responsible for importing `CustomLib` and creating the GraphQL client.

`AdminMFE` also uses `CustomLib` as a singleton library. However, it doesn't load `CustomLib` separately. Instead, it shares the instance of `CustomLib` that was already loaded by `HostMFE`. This is possible because of the singleton pattern implemented with webpack-module-federation. `AdminMFE` imports the GraphQL client that was created by `HostMFE` using `CustomLib` and uses it for its operations. I am rendering all users in admin page using usequery hook. This usequry hook uses graphlclient.

In summary, both `HostMFE` and `AdminMFE` share the same instance of `CustomLib` and the GraphQL client it creates, ensuring that the library and the client are not duplicated unnecessarily.

**Comment by jerelmiller** [comment]
@PrakashC1 if I'm reading this right, `CustomLib`, what you call "Library package.json", uses the `apollo-client` package and your MFEs `HostMFE` and `AdminMFE`, what you call "MFE package.json", uses `@apollo/client`. Is that correct?

Regardless, you'll want to make sure you're using `@apollo/client` in both places. `@apollo/client` is version 3 of the library and `apollo-client` is version 2 of the library. You can read more about the package changes in the [migration guide](https://www.apollographql.com/docs/react/migrating/apollo-client-3-migration).

The problem you're experiencing is that you're using `useQuery` using version 3 of the library with a client that is created using version 2 of the library. These are not compatible with each other since `useQuery` assumes it will be working with functionality present in the same version as itself.

Please upgrade everything that uses the `apollo-*` packages to use the equivalent v3 imports (`@apollo/client`) and this should resolve your issue.
